IRGen: Only emit PragmaCommentDecls if building for Windows, or LTO is enabled

slavapestov opened the pull request:
`clang::DeclContext::decls_begin()` deserializes all the pre-compiled headers
imported into the current translation unit, which is expensive.

Workaround this by disabling the emission of PragmaCommentDecls unless we're
building for a Windows target, or LTO is enabled.

A better fix would be to serialize a separate index for PragmaCommentDecls
so that they can be deserialized without deserializing everything else.
I filed rdar://problem/74036099 to track the longer-term fix.

Fixes rdar://problem/73951264.

slavapestov commented:
@martinboehme Does this look like a reasonable fix to you?

slavapestov commented:
Ok, this breaks the lto_autolink test unfortunately. @kateinoigakukun any ideas how we can get this information without deserializing everything?

kateinoigakukun commented:
> Ok, this breaks the lto_autolink test unfortunately. @kateinoigakukun any ideas how we can get this information without deserializing everything?

Sorry I have no idea how to get pragmas without loading all decls.

slavapestov committed:
IRGen: Only emit PragmaCommentDecls if building for Windows, or LTO is enabled

clang::DeclContext::decls_begin() deserializes all the pre-compiled headers
imported into the current translation unit, which is expensive.

Workaround this by disabling the emission of PragmaCommentDecls unless we're
building for a Windows target, or LTO is enabled.

A better fix would be to serialize a separate index for PragmaCommentDecls
so that they can be deserialized without deserializing everything else.
I filed rdar://problem/74036099 to track the longer-term fix.

Fixes rdar://problem/73951264.

slavapestov commented:
@kateinoigakukun @martinboehme I changed the fix to narrow it down to Windows and LTO. This makes the existing tests pass while recovering the performance regression in the common case.

slavapestov commented:
@swift-ci Please smoke test
